### PR TITLE
[master < T187-MAGE] All Shortest Paths Algorithm 

### PR DESCRIPTION
**Pull request by jmatak**
Implementation of `all shortest paths`. Things to consider: how to apply filtering of nodes when traversing paths, and how to include weighted paths along the way.

[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)

[Docs PR](https://github.com/memgraph/docs/pull/519)

### CODE REVIEW

File: licenses/BSL.txt [hunk 1]
```diff
@@ -36,7 +36,7 @@ ADDITIONAL USE GRANT:   You may use the Licensed Work in accordance with the
                         3. using the Licensed Work to create a work or solution
                         which competes (or might reasonably be expected to
                         compete) with the Licensed Work.
-CHANGE DATE:            2026-27-04
+CHANGE DATE:            2026-16-06
```
Comment (gitbuda) [inline licenses/BSL.txt:39]: This is changed on the release date, so it should be reverted here 😄

File: CMakeLists.txt [hunk 2]
```diff
@@ -54,7 +54,7 @@ option(MG_ENTERPRISE "Build Memgraph Enterprise Edition" ON)
 # Set the current version here to override the automatic version detection. The
 # version must be specified as `X.Y.Z`. Primarily used when building new patch
 # versions.
-set(MEMGRAPH_OVERRIDE_VERSION "")
+set(MEMGRAPH_OVERRIDE_VERSION "2.3.1")
```
Comment (gitbuda) [inline CMakeLists.txt:57]: Revert this as well

File: src/query/plan/operator.cpp [hunk 3]
```diff
@@ -1690,6 +1691,319 @@ class ExpandWeightedShortestPathCursor : public query::plan::Cursor {
...
+      };
+
+      const auto is_valid_duration = [&] {
+        return current_weight.IsDuration() && (current_weight >= TypedValue(utils::Duration(0), memory)).ValueBool();
+      };
+
+      if (!is_valid_numeric() && !is_valid_duration()) {
+        throw QueryRuntimeException("Calculated weight must be non-negative!");
+      }
+      //
```
Comment (jmatak) [inline src/query/plan/operator.cpp:1750]: This stuff is repeated in `wShortest` too.What is the best way to write a validation that checks multiple types and throws an exception? @antaljanosbenjamin @kostasrim

**Review by jmatak — COMMENTED** [review]
There are some things to comment out in the code itself. Otherwise, tests are written and the full feature is implemented. Moving PR to Ready for Review and starting documentation work.

File: CMakeLists.txt (hunk 2 above)
Comment (jmatak) [reply CMakeLists.txt]: Reverted, please resolve

File: src/query/plan/operator.cpp [hunk 4]
```diff
@@ -1690,6 +1691,319 @@ class ExpandWeightedShortestPathCursor : public query::plan::Cursor {
...
+      if (self_.common_.direction != EdgeAtom::Direction::OUT) {
+        auto in_edges = UnwrapEdgesResult(vertex.InEdges(storage::View::OLD, self_.common_.edge_types));
+        for (const auto &edge : in_edges) {
+          expand_vertex(edge, EdgeAtom::Direction::IN, weight, depth);
+        }
+      }
+    };
+
+    // Checkf if upper bound exists
+    if (self_.upper_bound_) {
```
Comment (BorisTasevski) [inline src/query/plan/operator.cpp:1802]: maybe replace this with tertiary

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

**Review by BorisTasevski — APPROVED** [review]

**Review by antepusic — APPROVED** [review]

File: src/query/frontend/ast/cypher_main_visitor.cpp [hunk 5]
```diff
@@ -1642,7 +1643,7 @@ antlrcpp::Any CypherMainVisitor::visitRelationshipPattern(MemgraphCypher::Relati
     };
     switch (relationshipLambdas.size()) {
       case 0:
-        if (edge->type_ == EdgeAtom::Type::WEIGHTED_SHORTEST_PATH)
+        if (edge->type_ == EdgeAtom::Type::WEIGHTED_SHORTEST_PATH || edge->type_ == EdgeAtom::Type::ALL_SHORTEST_PATHS)
```
Comment (antaljanosbenjamin) [inline src/query/frontend/ast/cypher_main_visitor.cpp:1646]: Please extend the error message below.

File: src/query/plan/operator.cpp [hunk 6]
```diff
@@ -787,7 +788,7 @@ ExpandVariable::ExpandVariable(const std::shared_ptr<LogicalOperator> &input, Sy
       weight_lambda_(weight_lambda),
       total_weight_(total_weight) {
   DMG_ASSERT(type_ == EdgeAtom::Type::DEPTH_FIRST || type_ == EdgeAtom::Type::BREADTH_FIRST ||
-                 type_ == EdgeAtom::Type::WEIGHTED_SHORTEST_PATH,
+                 type_ == EdgeAtom::Type::WEIGHTED_SHORTEST_PATH || type_ == EdgeAtom::Type::ALL_SHORTEST_PATHS,
```
Comment (antaljanosbenjamin) [inline src/query/plan/operator.cpp:791]: Please update the error message below.

File: src/query/plan/operator.cpp (hunk 3 above)
Comment (antaljanosbenjamin) [reply src/query/plan/operator.cpp:1750]: Just a single function that receives `current_weight` as parameter?

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: src/query/plan/operator.cpp (hunk 3 above)
Comment (jmatak) [reply src/query/plan/operator.cpp:1750]: `void` function that doesn't throw an exception once the input is valid?

**Review by antaljanosbenjamin — APPROVED** [review]
LGTM

**Pull request merged**